File: javascript/src/local-functions.ts
import type { FunctionCallOutputInput } from "./types/input.js";
import type { AgentResponse, FunctionCallOutputItem } from "./types/responses.js";

/** Pending client-executed function calls from a requires_action response. */
export function pendingFunctionCalls(response: AgentResponse): FunctionCallOutputItem[] {
  if (response.status !== "requires_action") {
    return [];
  }
  return response.output.filter((item): item is FunctionCallOutputItem => item.type === "function_call");
}

export function functionCallOutputInput(
  callID: string,
  output: string | Record<string, unknown>,
): FunctionCallOutputInput {
  return {
    type: "function_call_output",
    call_id: callID,
    output: typeof output === "string" ? output : JSON.stringify(output),
  };
}

export type LocalFunctionHandler = (args: Record<string, unknown>) => string | Record<string, unknown> | Promise<string | Record<string, unknown>>;

export type LocalFunctionHandlers = Record<string, LocalFunctionHandler>;

/** Run local handlers for pending function calls and return function_call_output input items. */
export async function runLocalFunctionHandlers(
  response: AgentResponse,
  handlers: LocalFunctionHandlers,
): Promise<FunctionCallOutputInput[]> {
  const pending = pendingFunctionCalls(response);
  const outputs: FunctionCallOutputInput[] = [];
  for (const call of pending) {
    const handler = handlers[call.name];
    if (!handler) {
      throw new Error(`no local handler registered for function ${call.name}`);
    }
    let args: Record<string, unknown> = {};
    if (call.arguments) {
      args = JSON.parse(call.arguments) as Record<string, unknown>;
    }
    const result = await handler(args);
    outputs.push(functionCallOutputInput(call.call_id, result));
  }
  return outputs;
}

File: python/src/agent_api/local_functions.py
from __future__ import annotations

import json
from collections.abc import Awaitable, Callable, Mapping
from typing import Any

from agent_api.types import AgentResponse, FunctionCallOutputInput

LocalFunctionHandler = Callable[[dict[str, Any]], str | dict[str, Any] | Awaitable[str | dict[str, Any]]]


def pending_function_calls(response: AgentResponse) -> list[dict[str, Any]]:
    if response.get("status") != "requires_action":
        return []
    return [item for item in response.get("output", []) if item.get("type") == "function_call"]


def function_call_output_input(call_id: str, output: str | dict[str, Any]) -> FunctionCallOutputInput:
    text = output if isinstance(output, str) else json.dumps(output)
    return {"type": "function_call_output", "call_id": call_id, "output": text}


async def run_local_function_handlers(
    response: AgentResponse,
    handlers: Mapping[str, LocalFunctionHandler],
) -> list[FunctionCallOutputInput]:
    pending = pending_function_calls(response)
    outputs: list[FunctionCallOutputInput] = []
    for call in pending:
        name = str(call.get("name", ""))
        handler = handlers.get(name)
        if handler is None:
            raise ValueError(f"no local handler registered for function {name}")
        raw_args = call.get("arguments") or "{}"
        args = json.loads(raw_args) if isinstance(raw_args, str) else dict(raw_args)
        result = handler(args)
        if hasattr(result, "__await__"):
            result = await result  # type: ignore[misc]
        call_id = str(call.get("call_id", ""))
        outputs.append(function_call_output_input(call_id, result))  # type: ignore[arg-type]
    return outputs
